Merge report sections (#68)
chhoumann/p4: Report/sections/2_design/operational_semantics.tex

	\begin{tabular}{l l}
		$a \in \textbf{ArithmExp}$ & $\text{Arithmetic expressions}$ \\ 
		$n \in \textbf{Num}$ & $\text{Numerals}$ \\
		$k \in \textbf{Var}$ & $\text{Variables}$ \\
		$A_v \in \textbf{AssignV}$ & $\text{Variable assignments}$ \\
		$DV \in \textbf{DeclV}$ & $\text{Variable declarations}$ \\
		$C_m \in \textbf{ContDecl}$ & $\text{Content declarations}$ \\
		$S \in \textbf{Stm}$ & $\text{Statements}$\\
		$DR \in \textbf{DecR}$ & $\text{Record declarations}$\\
		$r_n \in \textbf{Rnames}$ & $\text{Record names}$ \\
		$K \in \textbf{Gvar}$ & $\text{Generalized variables}$\\
		$P \in \textbf{Gpnames}$ & $\text{Generalized procedure names}$
	\end{tabular}

	\vspace{4mm}
	\textit{Formation rules}
	\begin{align*}
		K&::=k\mid r.K \\
		P&::=p\mid r.P \\

> 		S&::=K:=x=a\mid \texttt{skip}\mid S_1;S_2\mid \texttt{repeat }\{S\}; \mid L_\text{BRACK} \; D_V \; D_P \; D_R \; S \; R_\text{BRACK}\mid call \; P\\

Reviewer: ```suggestion
		S&::=K:=x=a\mid \texttt{skip}\mid S_1;S_2\mid \texttt{repeat }L_\text{BRACK} S R_\text{BRACK}; \mid L_\text{BRACK} \; D_V \; D_P \; D_R \; S \; R_\text{BRACK}\mid call \; P\\
```

		a&::=n\mid x\mid a_1+a_2\mid a_1-a_2\mid a_1*a_2\mid a_1/a_2\mid (a_1) \\
		%VList&::=[a]\mid [a_1, \ldots, a_n]\mid [\epsilon] \\
		D_V&::= x:=a; \; D_V \mid \epsilon \\

> 		D_P&::= p \; S; \; D_P \mid \epsilon \\

Reviewer: ```suggestion
		D_P&::= P \; S; \; D_P \mid \epsilon \\
```

Reviewer: Or was this intentional? Then we need a definition for p in the syntactic categories

> 		D_R&::= record \; r \; L_\text{BRACK} \; D_V \; D_P \; D_R \; R_\text{BRACK}; \; D_R \mid \epsilon

Reviewer: If some records can only contain other records, and the remaining records can only contain generalized variables and generalized procedures, should we then differentiate between these? We are unsure how we would do this. Would another formation rule do?

Reviewer: Also:

```suggestion
		D_R&::= record \; r \; L_\text{BRACK} \; D_V \; D_P \; D_R \; R_\text{BRACK} \; D_R \mid \epsilon
```

\begin{figure}[h]
	\centering
	\begin{tabular}{l l}
		$Env_R = Rnames \hookrightarrow Env_V \; x \; Env_P \; x \; Env_R$
	\end{tabular}
	\caption{Procedure environment}
	\label{fig:EnvRDef}
\end{figure}

The definition of \figref{EnvRDef} ensures that a record environment has the information of variables, procedures and other records bound within a given record.

Similarly, to ensure that every procedure is able to obtain the records, procedures and variables when a procedure is declared we give the definition as seen in \figref{EnvPDef}:

\begin{figure}[h]
	\centering
	\begin{tabular}{l l}
		$Env_P = Pnames \hookrightarrow Env_V \; x \; Env_P \; x \; Env_R$
	\end{tabular}
	\caption{Procedure environment}
	\label{fig:EnvPDef}
\end{figure}

With these definitions we ensure that we have static scope rules as we introduce the transition rules for \dazel{}.

As well, because of the object structure of \dazel{} mentioned previously, we have to introduce semantics for variables and procedures, such that it is possible to describe the fact that they appear inside of objects as well as their location. For this we introduce generalized variables and generalized procedure names.

> Generalized variables \textbf{Gvar} are sequences of record names r1..rn.k concluded by a variable name k. The range over \textbf{Gvar} is caputered by \textit{K}.  
> Similarly, generalized procedure names are a sequence of procedures r1..rn.p that range over \textbf{Gpnames} captured by \textit{P}.

Reviewer: Remember to fix subscripts
```suggestion
Generalized variables \textbf{Gvar} are sequences of record names $r_1 \ldots r_n.k$ concluded by a variable name k. The range over \textbf{Gvar} is caputered by \textit{K}.  
Similarly, generalized procedure names are a sequence of procedures $r_1 \ldots r_n.p$ that range over \textbf{Gpnames} captured by \textit{P}.
```

Reviewer: And source - as we talked about

		\qquad \text{ where }v = v_1 \cdot v_2
		\\
		% 
		\\
		[DIVIDE_\text{BSS}] 
		\qquad \dfrac
		{env_R, \envs \vdash a_1 \rightarrow_a v_1 \; env_R, \envs \vdash a_2 \rightarrow_a v_2}
		{env_R, \envs \vdash a_1 / a_2 \rightarrow_a v}
		\qquad \text{ where }v = \dfrac{v_1}{v_2}
		\\
		% 
		\\
		[PARENTHESES_\text{BSS}] 
		\qquad \dfrac
		{env_R, \envs \vdash a_1 \rightarrow_a v_1}
		{env_R, \envs \vdash (a_1) \rightarrow_a v_1}
		\\
		% 
		\\
		[NUM_\text{BSS}] 
		\qquad env_R, envs \vdash n \rightarrow_a v
		\qquad \text{ if } \mathbb{R}[\![n]\!] = v \quad \text{Where } \mathcal{R}:\textbf{Num}\rightarrow \mathbb{R}
		\\
		% 
		\\
	\end{gather*}
	\caption{Big-step transition rules for \textbf{Arithmetic expressions}.}
	\label{fig:BssArithm}
\end{figure}

\begin{figure}[htbp]
	\centering
	\begin{gather*}
		[VAR-DECL_\text{BSS}]
		\qquad 
		\dfrac
		{env_R \vdash \langle D_V,\; env'_V, \; sto[l \mapsto v] \rangle \rightarrow_\text{DV} (env'_V, sto')}
		{env_R \vdash \langle x:=a; D_V, env_V, sto \rangle \rightarrow_\text{DV} (env'_V, sto')} \\
		\text{where } \textit{env}_R, \textit{env}_V, \textit{sto} \vdash \text{a} \rightarrow_a \text{v} \\

> 		\text{and} \textit{l} = \textit{env}_V \; \text{next} \\
> 		\text{and} \textit{env'}_V = \textit{env}_V[k \mapsto l][next \mapsto new l]

Reviewer: Spacing

```suggestion
		\text{and } \textit{l} = \textit{env}_V \; \text{next} \\
		\text{and } \textit{env'}_V = \textit{env}_V[k \mapsto l][next \mapsto \textit{ new } l]
```

		\\
	\end{gather*}
	\caption{Big-step transition rules for \textbf{generalized variable declarations}.}
	\label{fig:BssStm}
\end{figure}

\begin{figure}[htbp]
	\centering
	\begin{gather*}
		[MEMBERACCESS_\text{BSS}]
		\qquad
		\dfrac{Env_R, Env_V \vdash \langle  }{}	
	\end{gather*}
	\caption{}
	\label{fig:BssStm}
\end{figure}

\begin{figure}[htbp]
	\centering
	\begin{gather*}
		[ASSIGNMENT_\text{BSS}] 
		\qquad env_R, env_V, env_P \vdash \langle K:=a,sto\rangle \rightarrow sto[l\mapsto v] \\
		\qquad \text{ where } env_R, env_V, sto \vdash a \rightarrow_a v\\
		\qquad \text{ and } env_R, env_V \vdash K \rightarrow \textit{l}
		\\
		% 
		\\
		[COMPOUND_\text{BSS}] 
		\qquad \dfrac
		{env_R, env_V, env_P \vdash \langle S_1, sto \rangle \rightarrow sto'' \; env_R, env_V, env_P \vdash \langle S_2, sto'' \rangle \rightarrow sto'}
		{env_R, env_V, env_P \vdash \langle S_1;S_2,sto \rangle\rightarrow sto'}
		\\
		% 
		\\
		[REPEAT_\text{BSS}] 
		\qquad \dfrac
		{\langle S_1, sto \rangle \rightarrow sto'' \;\langle \texttt{repeat} \{S\}, sto \rangle \rightarrow sto'}
		{\langle S_1 \texttt{;repeat} \{S\}, sto \rangle\rightarrow sto'}
		\\
		% 
		\\
		[Block_\text{BSS}] 
		\qquad \frac
		{

> 			\begin{gathered}
> 				env_R \vdash \langle D_V, env_V, sto \rangle \rightarrow_DV (env'_V, sto'')
> 				env'_V \vdash \langle D_R, env_R \rangle \rightarrow_DR (env'_R, env''_V)
> 				env''_V \vdash \langle D_P, env_P \rangle \rightarrow_DP env'_P
> 				env'_R, env''_V, env'_p \vdash \langle S, sto'' \rangle \rightarrow sto'
> 			\end{gathered}

Reviewer: ```suggestion
			\begin{gathered}
				env_R \vdash \langle D_V, env_V, sto \rangle \rightarrow_DV (env'_V, sto'') \\
				env'_V \vdash \langle D_R, env_R \rangle \rightarrow_DR (env'_R, env''_V) \\
				env''_V \vdash \langle D_P, env_P \rangle \rightarrow_DP env'_P \\
				env'_R, env''_V, env'_p \vdash \langle S, sto'' \rangle \rightarrow sto'
			\end{gathered}
```

		}
		{env_R, env_V, env_P \vdash \langle L_\text{Brace} \; D_V \; D_P \; S \; R_\text{Brace}, \; sto\rangle\rightarrow sto'}
		\\
		% 
		\\
		[CALL_\text{BSS}]

> 		\qquad \dfrac{env'_R, env'_V, env'_p \vdash \langle S,sto \rangle \rightarrow sto'}{env_R, env_V, env_P \vdash \langle call P, sto \rightarrow sto'}

Reviewer: ```suggestion
		\qquad \dfrac{env'_R, env'_V, env'_p \vdash \langle S,sto \rangle \rightarrow sto'}{env_R, env_V, env_P \vdash \langle \textit{ call } P, sto \rightarrow sto'}
```
